NativeAOT Fully implements warning suppressions logic

vitek-karas opened the pull request:
Ports the warning suppression logic from linker. The only unsupported feature is reading the suppressions from attribute XML. This also doesn't port the "redundant suppression detection" logic from linker.

Other changes:
* Extracts some helpers for attribute handling into CustomAttributeExtensions (mostly around properties and events)
* Implements equality and hashcode on property and event descriptors - so that they can be used as keys in a dictionary
* Implements passing along the "--singlewarn" command line option in the test infra.

Also ports over all of the warning suppression tests from linker:
* Modified all of the "redundant suppressions" tests to be "trimmer-only" as that feature is only implemented there.
* Removed the "suppressions from XML" tests

Per discussion with @MichalStrehovsky I will rename `PropertyPseudoDesc` and `EventPseudoDesc` to `EcmaProperty` and `EcmaEvent`. But I'll do that in a separate PR as it would add a lot of noise to this one.

Fixes https://github.com/dotnet/runtime/issues/81129

vitek-karas committed:
First

vitek-karas committed:
Progress

vitek-karas committed:
Everything passes

vitek-karas committed:
Cleanup

tlakollo reviewed (approved):
Looks good to me

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CustomAttributeExtensions.cs:
```diff
@@ -0,0 +1,121 @@
+                    return property.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case EventPseudoDesc @event:
+                    return @event.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case ModuleDesc module:
+                    if (module is not EcmaAssembly ecmaAssembly)
+                        return Enumerable.Empty<CustomAttributeValue<TypeDesc>>();
+                    return ecmaAssembly.GetDecodedCustomAttributesForModule(attributeNamespace, attributeName)
+                        .Concat(ecmaAssembly.GetDecodedCustomAttributes(attributeNamespace, attributeName));
```
Combining assembly and module attributes is _very subtle_.

I'd prefer this whole helper to be more local to logging - e.g. stash it in a ILCompiler.Logging.Extensions namespace that only logging implementation will have a `using` for.

We try to be more deliberate on what is accepted by the compiler in general.

Also we try to avoid allocating. Putting the allocating APIs into the global `ILCompiler` namespace will guide people towards the wrong APIs. It's fine to use APIs like this when generating warnings. There won't be millions of those. But the rest of the compiler should keep using the less convenient APIs.

vitek-karas replied:
I was struggling with this... so thanks for some guidance. I will keep the property/event/module methods in the general extensions class (as they simply supplement the ones available from the type system already). And I will move the "get attributes from anything" to the warning suppression class alone. I'll get rid of the module/assembly handling completely and to it in the one place where it's actually used.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Logging/UnconditionalSuppressMessageAttributeState.cs:
```diff
@@ -0,0 +1,338 @@
+
+        public UnconditionalSuppressMessageAttributeState(CompilerGeneratedState? compilerGeneratedState, Logger logger)
+        {
+            _suppressions = new Dictionary<TypeSystemEntity, Dictionary<int, Suppression>>();
```
Do we have any thread safety concerns?

vitek-karas replied:
Great point - sorry, still in the single-threaded illink mindset.

vitek-karas committed:
PR feedback

The biggest change is to use lock free hashtables instead of a dictionary to avoid races.

Note on that - had to avoid the typical GetOrCreateValue because of possible recursion. And also we don't want to store empty records, it's actually very fast to figure out that an entity doesn't have suppressions from the type system, no point in wasting memory to remember that result.aot-run

vitek-karas commented:
@MichalStrehovsky  could you please take another look - specifically at the thread-safe implementation now.

MichalStrehovsky reviewed (approved):
Looks reasonable.

Although I do wonder whether the cache is a premature optimization - we need to look at suppressions if there's a warning. It doesn't feel very expensive to just walk up the hierarchy and look at all the suppression attributes when we hit a warning. Most places will not have such attribute.

vitek-karas committed:
Avoid caching suppressions to simplify code

After discussing this we think it's not worth the code complexity to cache the suppressions. So this version will always do all of the attribute lookups and parsing on each warning.

Note that it still needs one tiny cache - a hashtable of assemblies, to avoid reporting the warnings produced by suppression parsing more than once.

The code is a bit more complex since I tried to avoid allocating things unnecessarily - there's probably more which can be done in this regard though.

vitek-karas commented:
New version which avoids caches.

MichalStrehovsky reviewed (approved):
Thank you! Note there's a build break

vitek-karas committed:
Fix build
